File: modules/widgets/lib/aerox/widgets/containers/Fitter.cpp
﻿#include "aerox/widgets/containers/Fitter.hpp"

#include "aerox/widgets/ContainerSlot.hpp"

namespace aerox::widgets
{
    FitMode Fitter::GetMode() const
    {
        return _mode;
    }

    void Fitter::SetMode(FitMode mode)
    {
        _mode = mode;
    }

    Vec2<float> Fitter::ComputeDesiredSize()
    {
        if(auto slot = GetSlot(0))
        {
            return slot->GetWidget()->GetDesiredSize();
        }

        return Vec2{0.0f};
    }

    size_t Fitter::GetMaxSlots() const
    {
        return 1;
    }

    void Fitter::ArrangeSlots(const Vec2<float>& drawSize)
    {
        SizeContent(drawSize);
    }

    Vec2<float> Fitter::ComputeContainSize(const Vec2<float>& drawSize, const Vec2<float>& widgetSize)
    {
        auto aspect = widgetSize.y / widgetSize.x;
        Vec2 scaledWidgetSize{drawSize.x,drawSize.x * aspect};

        if(drawSize.NearlyEquals(scaledWidgetSize,0.001)) return scaledWidgetSize;

        return scaledWidgetSize.y < drawSize.y  ? scaledWidgetSize : Vec2{drawSize.y / aspect,drawSize.y};
    }

    Vec2<float> Fitter::ComputeCoverSize(const Vec2<float>& drawSize, const Vec2<float>& widgetSize)
    {
        auto aspect = widgetSize.y / widgetSize.x;
        Vec2 scaledWidgetSize{drawSize.x,drawSize.x * aspect};

        if(drawSize.NearlyEquals(scaledWidgetSize,0.001)) return scaledWidgetSize;

        return scaledWidgetSize.y < drawSize.y  ? Vec2{drawSize.y / aspect,drawSize.y} : scaledWidgetSize;
    }

    void Fitter::SizeContent(const Vec2<float>& size) const
    {
        if(auto slot = GetSlot(0))
        {
            auto widget = slot->GetWidget();
            auto widgetSize = widget->GetDesiredSize();
            auto newDrawSize = widgetSize;
            if(!widgetSize.NearlyEquals(size,0.001))
            {
                switch (GetMode())
                {
                case FitMode::None:
                    break;
                case FitMode::Fill:
                    {
                        newDrawSize = size;
                    }
                    break;
                case FitMode::Contain:
                    {
                        newDrawSize = ComputeContainSize(widgetSize,size);
                    }
                    break;
                case FitMode::Cover:
                    {
                        newDrawSize = ComputeCoverSize(widgetSize,size);
                    }
                    break;
                }
            }

            if(!widgetSize.NearlyEquals(newDrawSize,0.001)) widget->SetDrawSize(newDrawSize);
            
            auto halfSelfDrawSize = size / 2.0f;
            auto halfSlotDrawSize = newDrawSize / 2.0f;

            auto diff = halfSelfDrawSize - halfSlotDrawSize;

            if(!widget->GetRelativeOffset().NearlyEquals(diff,0.001)) widget->SetRelativeOffset(diff);
        }
    }
}

File: rin.Framework/Views/Composite/Fitter.cs
﻿using rin.Framework.Core.Math;
using rin.Framework.Views.Enums;
using rin.Framework.Views.Graphics;
using rin.Framework.Views.Layouts;

namespace rin.Framework.Views.Composite;

public enum FitMode
{
    Fill,
    Contain,
    Cover,
    None
}

/// <summary>
/// </summary>
public class Fitter : SingleSlotCompositeView
{

    private FitMode _fitFittingMode = FitMode.Fill;
    /// <summary>
    /// Adds the View to this container
    /// </summary>
    
    public Fitter() : base()
    {
        Clip = Clip.Bounds;
    }

    public FitMode FittingMode
    {
        get => _fitFittingMode;
        set
        {
            var old = _fitFittingMode;
            _fitFittingMode = value;
            if (_fitFittingMode != old)
            {
                FitContent(GetContentSize());
            }
        }
    }
    

    protected override Vector2<float> ComputeDesiredContentSize()
    {
        if (GetSlot() is { } slot)
        {
            return slot.Child.GetDesiredSize();
        }

        return 0.0f;
    }

    public static Vector2<float> ComputeContainSize(Vector2<float> drawSize, Vector2<float> widgetSize)
    {
        var widgetAspect = widgetSize.Y / widgetSize.X;
        var scaledWidgetSize = new Vector2<float>(drawSize.X, drawSize.X * widgetAspect);

        if (drawSize.Equals(scaledWidgetSize)) return scaledWidgetSize;

        return scaledWidgetSize.Y <= drawSize.Y
            ? scaledWidgetSize
            : new Vector2<float>(drawSize.Y / widgetAspect, drawSize.Y);
    }

    public static Vector2<float> ComputeCoverSize(Vector2<float> drawSize, Vector2<float> widgetSize)
    {
        var widgetAspect = widgetSize.Y / widgetSize.X;
        var scaledWidgetSize = new Vector2<float>(drawSize.X, drawSize.X * widgetAspect);

        if (drawSize.Equals(scaledWidgetSize)) return scaledWidgetSize;


        return scaledWidgetSize.Y <= drawSize.Y
            ? new Vector2<float>(drawSize.Y / widgetAspect, drawSize.Y)
            : scaledWidgetSize;
    }

    public Vector2<float> FitContent(Vector2<float> drawSize)
    {
        if (GetSlot() is { } slot)
        {
            var widget = slot.Child;
            var widgetSize = widget.GetDesiredSize();
            var newDrawSize = _fitFittingMode switch
                {
                    FitMode.Fill => drawSize,
                    FitMode.Contain => ComputeContainSize(drawSize, widgetSize),
                    FitMode.Cover => ComputeCoverSize(drawSize, widgetSize),
                    FitMode.None => widgetSize,
                    _ => throw new ArgumentOutOfRangeException()
                };

            widget.ComputeSize(newDrawSize);


            var halfSelfDrawSize = drawSize;
            halfSelfDrawSize /= 2.0f;
            var halfSlotDrawSize = newDrawSize;
            halfSlotDrawSize /= 2.0f;

            var diff = halfSelfDrawSize - halfSlotDrawSize;

            widget.Offset = diff;

            return drawSize;
        }

        return drawSize;
    }

    protected override Vector2<float> ArrangeContent(Vector2<float> availableSpace)
    {
        var desired = GetDesiredContentSize();
        return FitContent(new Vector2<float>(float.IsFinite(availableSpace.X) ? availableSpace.X : desired.X,
            float.IsFinite(availableSpace.Y) ? availableSpace.Y : desired.Y));
    }

    public override void OnChildInvalidated(View child, InvalidationType invalidation)
    {
        FitContent(GetContentSize());
    }

    public override IEnumerable<ISlot> GetSlots()
    {
        if (GetSlot() is { } slot)
        {
            return [slot];
        }

        return [];
    }
}